### PR TITLE
cmake/kokkos_tribits.cmake: add parsing for args

### PR DESCRIPTION
**Pull request by ndellingwood**
Update the `KOKKOS_ADD_EXECUTABLE_AND_TEST` function to parse args.
Changes motivated by failures of `UnitTest_CMakePassCmdLineArgs` when testing VOTD Kokkos with Trilinos, failure was due to all arguments not reaching the test causing throw "Up". These changes resolved the issue for me.

### CODE REVIEW

File: cmake/kokkos_tribits.cmake [hunk 1]
```diff
@@ -136,13 +136,22 @@ ENDFUNCTION()
 
 FUNCTION(KOKKOS_ADD_EXECUTABLE_AND_TEST ROOT_NAME)
 IF (KOKKOS_HAS_TRILINOS)
+  CMAKE_PARSE_ARGUMENTS(PARSE
+    ""
+    ""
+    "SOURCES;CATEGORIES;ARGS"
+    ${ARGN})
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake:143]: We are now parsing the arguments in both branches. Let's move it before the if-switch (and also add `VERIFY_EMPTY(KOKKOS_ADD_EXECUTABLE_AND_TEST ${PARSE_UNPARSED_ARGUMENTS})` check).

Comment (jjwilke) [reply cmake/kokkos_tribits.cmake:143]: Yeah, good call.

File: cmake/kokkos_tribits.cmake [hunk 2]
```diff
@@ -136,13 +136,22 @@ ENDFUNCTION()
...
+  IF(DEFINED PARSE_ARGS)
+    STRING(REPLACE ";" " " PARSE_ARGS "${PARSE_ARGS}")
```
Comment (jjwilke) [inline cmake/kokkos_tribits.cmake:145]: I don't think this is necessary here? You should be able to pass this through directly as a list. Were you getting some sort of error?

**Review by jjwilke — DISMISSED** [review]

File: cmake/kokkos_tribits.cmake (hunk 1 above)
Comment (ndellingwood) [reply cmake/kokkos_tribits.cmake]: I moved the parsing outside of the if/else and remove the duplication, thanks.
The `VERIFY_EMPTY` check was in `fake_tribits` and was not found when I tried to call outside of the `ELSE` branch (that is within the `NOT KOKKOS_HAS_TRILINOS` branch), so I moved the `VERIFY_EMPTY` function to `kokkos_tribits`

File: cmake/kokkos_tribits.cmake (hunk 2 above)
Comment (ndellingwood) [reply cmake/kokkos_tribits.cmake:152]: This was necessary to make sure all three command-line arguments for the `UnitTest_CMakePassCmdLineArgs` test were properly passed to it by TriBITS; without this, 3 separate unit tests were created, one per argument, with each test failing (the unit test verifies all three arguments as the passing condition).

I was confused by the problem but did a check on the TriBITS documentation and it says this about multiple arguments, I'm guessing this is why this change resolved the issue?

> WARNING: Multiple arguments passed to a single test invocation must be quoted or multiple tests taking single arguments will be created instead! 

https://tribits.org/doc/TribitsDevelopersGuide.html#tribits-add-test

**Comment by ndellingwood** [comment]
@jjwilke @masterleinad once any further needed changes are complete I'll rebase to a single commit, thanks for feedback!

File: cmake/kokkos_tribits.cmake [hunk 3]
```diff
@@ -6,6 +6,12 @@ INCLUDE(GNUInstallDirs)
 
 MESSAGE(STATUS "The project name is: ${PROJECT_NAME}")
 
+FUNCTION(VERIFY_EMPTY CONTEXT)
+if(${ARGN})
+MESSAGE(FATAL_ERROR "Kokkos does not support all of Tribits. Unhandled arguments in ${CONTEXT}:\n${ARGN}")
+endif()
+ENDFUNCTION()
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake:13]: I know that it's pre-existing but we could still improve the indentation
```suggestion
FUNCTION(VERIFY_EMPTY CONTEXT)
  if(${ARGN})
    MESSAGE(FATAL_ERROR "Kokkos does not support all of Tribits. Unhandled arguments in ${CONTEXT}:\n${ARGN}")
  endif()
ENDFUNCTION()
```

File: cmake/kokkos_tribits.cmake [hunk 4]
```diff
@@ -135,22 +141,26 @@ FUNCTION(KOKKOS_ADD_EXECUTABLE ROOT_NAME)
 ENDFUNCTION()
 
 FUNCTION(KOKKOS_ADD_EXECUTABLE_AND_TEST ROOT_NAME)
+CMAKE_PARSE_ARGUMENTS(PARSE
+  ""
+  ""
+  "SOURCES;CATEGORIES;ARGS"
+  ${ARGN})
+  VERIFY_EMPTY(KOKKOS_ADD_EXECUTABLE_AND_TEST ${PARSE_UNPARSED_ARGUMENTS})
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake:149]:
```suggestion
VERIFY_EMPTY(KOKKOS_ADD_EXECUTABLE_AND_TEST ${PARSE_UNPARSED_ARGUMENTS})
```

**Review by masterleinad — COMMENTED** [review]
Just some indentation suggestions. Otherwise, this looks good to me.

File: cmake/kokkos_tribits.cmake (hunk 3 above)
Comment (ndellingwood) [reply cmake/kokkos_tribits.cmake]: Addressed by SHA 84cbc1b , thanks!

File: cmake/kokkos_tribits.cmake (hunk 4 above)
Comment (ndellingwood) [reply cmake/kokkos_tribits.cmake]: Addressed by SHA 84cbc1b , thanks!

File: cmake/kokkos_tribits.cmake (hunk 2 above)
Comment (jjwilke) [reply cmake/kokkos_tribits.cmake:152]: Oh, gotcha. I had misunderstood Tribits here.  In that case, we want the function behavior to match TriBITS. We don't want the kokkos wrappers to have different behavior.  We need to change the unit_test/CMakeLists.txt to pass the args as a single string.

**Review by masterleinad — APPROVED** [review]

File: cmake/kokkos_tribits.cmake (hunk 2 above)
Comment (ndellingwood) [reply cmake/kokkos_tribits.cmake:152]: @jjwilke I tried removing the `STRING(REPLACE ";" " " PARSE_ARGS "${PARSE_ARGS}")` and adding quotes around the args for the UnitTest_CMakePassCmdLineArgs.cpp test in core/unit_test/CMakeLists.txt (i.e. `ARGS "one 2 THREE"`) but running into some whack-a-mole fun; this still passes in Trilinos as expected but fails in a standalone Kokkos build. Any suggestions?

**Comment by ndellingwood** [comment]
CI had OMPTarget-clang error unrelated to PR
```
OCI runtime create failed: container_linux.go:349: starting container process caused "process_linux.go:449: container init caused \"process_linux.go:432: running prestart hook 1 caused \\\"error running hook: exit status 1, stdout: , stderr: nvidia-container-cli: initialization error: nvml error: driver not loaded\\\\n\\\"\"": unknown
```
retest this please

File: cmake/kokkos_tribits.cmake (hunk 2 above)
Comment (jjwilke) [reply cmake/kokkos_tribits.cmake:153]: Yeah, my bad. I was misunderstanding what TriBITS wanted. I pushed a change that hopefully fixes things.

Comment (ndellingwood) [reply cmake/kokkos_tribits.cmake:153]: @jjwilke I tested the standalone Kokkos build and I'm seeing configure errors like this:

```
CMake Error at cmake/fake_tribits.cmake:139 (SET_TESTS_PROPERTIES):
  SET_TESTS_PROPERTIES Can not find test to add properties to:
  KokkosCore_UnitTest_Serial1
Call Stack (most recent call first):
  cmake/kokkos_tribits.cmake:171 (KOKKOS_ADD_TEST)
  core/unit_test/CMakeLists.txt:174 (KOKKOS_ADD_EXECUTABLE_AND_TEST)
...#etc
```

File: cmake/fake_tribits.cmake [hunk 5]
```diff
@@ -106,7 +100,7 @@ FUNCTION(KOKKOS_ADD_TEST)
       ${TEST_UNPARSED_ARGUMENTS}
     )
 
-    if(TEST_TOOL)
+  if(TEST_TOOL)
```
Comment (masterleinad) [inline cmake/fake_tribits.cmake:103]: This looks wrong.
```suggestion
    if(TEST_TOOL)
```

File: core/unit_test/CMakeLists.txt [hunk 6]
```diff
@@ -539,7 +539,7 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
 KOKKOS_ADD_EXECUTABLE_AND_TEST(
   UnitTest_CMakePassCmdLineArgs
   SOURCES UnitTest_CMakePassCmdLineArgs.cpp
-  ARGS one 2 THREE
+  ARGS "one 2 THREE"
```
Comment (dalg24) [inline core/unit_test/CMakeLists.txt:542]: Why did you change this?

Comment (ndellingwood) [reply core/unit_test/CMakeLists.txt:542]: @dalg24 this was done intentionally to match TriBITS behavior, here's link to the discussion comment
https://github.com/kokkos/kokkos/pull/3457#discussion_r501203687

**Comment by ndellingwood** [comment]
I verified that with the most recent set of changes da465fd `UnitTest_CMakePassCmdLineArgs` now passes in a Trilinos build; CI+Travis confirm the changes work with standalone Kokkos.

**Review by masterleinad — COMMENTED** [review]
I checked manually that the new behavior matches the TriBITS description. I didn't check compiling and running with `Trilinos`, though.

File: core/unit_test/CMakeLists.txt (hunk 6 above)
Comment (jjwilke) [reply core/unit_test/CMakeLists.txt:542]: Yeah, this is a tribits issue. We have been following the rule that function `KOKKOS_X` has the exact same behavior as the `TRIBITS_X` function. I'd prefer not to break that rule.

Comment (dalg24) [reply core/unit_test/CMakeLists.txt:542]: But according to the link to the documentation that was posted earlier https://tribits.org/doc/TribitsDevelopersGuide.html#tribits-add-test this was valid syntax

Comment (jjwilke) [reply core/unit_test/CMakeLists.txt:542]: Sorry, didn't explain well. You can pass multiple argument strings in, but that generates multiple tests. You create a test for entry in the list.

File: cmake/fake_tribits.cmake [hunk 7]
```diff
@@ -127,24 +122,47 @@ FUNCTION(KOKKOS_ADD_TEST)
     # These should be the full target name
     SET(TEST_NAME ${PACKAGE_NAME}_${TEST_NAME})
     SET(EXE ${PACKAGE_NAME}_${EXE_ROOT})
-    IF(WIN32)
-      ADD_TEST(NAME ${TEST_NAME} WORKING_DIRECTORY ${LIBRARY_OUTPUT_PATH} COMMAND ${EXE}${CMAKE_EXECUTABLE_SUFFIX} ${TEST_ARGS})
+    IF (TEST_ARGS)
+      SET(TEST_NUMBER 0)
+      FOREACH (ARG_STR ${TEST_ARGS})
+        message("adding test with args ${ARG_STR}")
```
[the remark this answers is not in the document]
Comment (dalg24) [reply cmake/fake_tribits.cmake:128]: Of course I merged that...

File: cmake/fake_tribits.cmake [hunk 8]
```diff
@@ -127,24 +122,47 @@ FUNCTION(KOKKOS_ADD_TEST)
...
-    IF(TEST_WILL_FAIL)
-      SET_TESTS_PROPERTIES(${TEST_NAME} PROPERTIES WILL_FAIL ${TEST_WILL_FAIL})
-    ENDIF()
-    IF(TEST_FAIL_REGULAR_EXPRESSION)
-      SET_TESTS_PROPERTIES(${TEST_NAME} PROPERTIES FAIL_REGULAR_EXPRESSION ${TEST_FAIL_REGULAR_EXPRESSION})
-    ENDIF()
-    IF(TEST_PASS_REGULAR_EXPRESSION)
-      SET_TESTS_PROPERTIES(${TEST_NAME} PROPERTIES PASS_REGULAR_EXPRESSION ${TEST_PASS_REGULAR_EXPRESSION})
+      IF(WIN32)
+        ADD_TEST(NAME ${TEST_NAME}${TEST_NUMBER} WORKING_DIRECTORY ${LIBRARY_OUTPUT_PATH}
```
[the remark this answers is not in the document]
Comment (dalg24) [reply cmake/fake_tribits.cmake:143]: @jjwilke Please address this comment and the one below
